pydata_sphinx_theme/__init__.py

```diff
@@ -38,26 +67,59 @@ def get_nav_object(maxdepth=None, collapse=True, **kwargs):
         # The TocTree will contain the full site TocTree including sub-pages.
         # "collapse=True" collapses sub-pages of non-active TOC pages.
         # maxdepth controls how many TOC levels are returned
-        toctree = TocTree(app.env).get_toctree_for(
+        toc = MyTocTree(app.env)
+        toctree = toc.get_toctree_for(
             pagename, app.builder, collapse=collapse, maxdepth=maxdepth, **kwargs
         )
+
         # If no toctree is defined (AKA a single-page site), skip this
         if toctree is None:
             return []
 
+        if subpage_caption:
+            if pagename not in [app.env.config.master_doc, "genindex", "search"]:
+                def is_first_active_page(node):
+                    return isinstance(node, nodes.bullet_list) and node.attributes.get("iscurrent")
+
+                active_first_page = list(toctree.traverse(is_first_active_page))[0]
+                # A path to the active TOC item's first page, relative to the current page
+                first_page_path = list(active_first_page.traverse(nodes.reference))[0].attributes.get("refuri")
+                if first_page_path == "":
+                    # First TOC item's first page *is* the active page
+                    first_page_path = Path(pagename).name
+                else:
+                    first_page_path = Path(first_page_path).with_suffix("")
+                rel_first_page_path = str(Path(pagename).parent.joinpath(first_page_path))
+
+                # We only wish to show a single page's descendants, so we'll keep their captions
+                subpage_toctree = toc.get_toctree_for_subpage(
+                    rel_first_page_path, app.builder, collapse=collapse, maxdepth=maxdepth, **kwargs
+                )
+                if subpage_toctree is not None:
+                    # Find the current page in the top-level children
+                    for item in toctree.children:
+                        if isinstance(item, nodes.bullet_list) and item.attributes.get("iscurrent", []):
+                            # Append that pages' toctree so we get captions
+                            subpage_list = item.children[0]
+                            subpage_list.children = [subpage_list.children[0]] + subpage_toctree.children        
+
         # toctree has this structure
         #   <caption>
         #   <bullet_list>
         #       <list_item classes="toctree-l1">
         #       <list_item classes="toctree-l1">
         # `list_item`s are the actual TOC links and are the only thing we want
-        toc_items = [item for child in toctree.children for item in child
-                     if isinstance(item, docutils.nodes.list_item)]
+        toc_items = []
+        for child in toctree.children:
```

Comment:
a lot of this extra code is because we were making really strong assumptions about the structure of the toctree, so we need to handle some other edge cases